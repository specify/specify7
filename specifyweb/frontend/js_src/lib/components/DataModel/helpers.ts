import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
