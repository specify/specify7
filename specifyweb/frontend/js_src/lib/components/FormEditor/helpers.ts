import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
