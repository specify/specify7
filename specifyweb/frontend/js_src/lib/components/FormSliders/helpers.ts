import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
