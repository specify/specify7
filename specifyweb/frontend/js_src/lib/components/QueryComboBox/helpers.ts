import type { RA, WritableArray } from '../../utils/types';
import { toTable, toTreeTable } from '../DataModel/helpers';
import type { AnySchema } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { idFromUrl } from '../DataModel/resource';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { tables } from '../DataModel/tables';
import type { SpQuery, SpQueryField } from '../DataModel/types';
import { getMainTableFields } from '../Formatters/formatters';
import { userInformation } from '../InitialContext/userInformation';
import { userPreferences } from '../Preferences/userPreferences';
import { queryFieldFilters } from '../QueryBuilder/FieldFilter';
import { QueryFieldSpec } from '../QueryBuilder/fieldSpec';
import { flippedSortTypes } from '../QueryBuilder/helpers';
import type { TypeSearch } from './spec';
import type { CollectionRelationships } from './useCollectionRelationships';
import type { QueryComboBoxTreeData } from './useTreeData';

export function makeComboBoxQuery({
  fieldName,
  value,
  isTreeTable,
  typeSearch: { table },
  specialConditions,
}: {
  readonly fieldName: string;
  readonly value: string;
  readonly isTreeTable: boolean;
  readonly typeSearch: TypeSearch;
  readonly specialConditions: RA<SpecifyResource<SpQueryField>>;
}): SpecifyResource<SpQuery> {
  const query = new tables.SpQuery.Resource({}, { noBusinessRules: true });
  query.set('name', 'Ephemeral QueryCBX query');
  query.set('contextName', table.name);
  query.set('contextTableId', table.tableId);
  query.set('selectDistinct', false);
  query.set('countOnly', false);
  query.set('specifyUser', userInformation.resource_uri);
  query.set('isFavorite', false);
  query.set('ordinal', null);

  const searchAlgorithm = userPreferences.get(
    'form',
    'queryComboBox',
    isTreeTable ? 'treeSearchAlgorithm' : 'searchAlgorithm'
  );
  const searchField = QueryFieldSpec.fromPath(table.name, fieldName.split('.'))
    .toSpQueryField()
    .set('isDisplay', false)
    .set('startValue', searchAlgorithm === 'contains' ? `%${value}%` : value)
    .set(
      'operStart',
      searchAlgorithm === 'contains'
        ? queryFieldFilters.like.id
        : queryFieldFilters.startsWith.id
    );

  const displayField = QueryFieldSpec.fromPath(table.name, [])
    .toSpQueryField()
    .set('isDisplay', true)
    .set('sortType', flippedSortTypes.ascending);

  query.set('fields', [searchField, displayField, ...specialConditions]);

  return query;
}

export function getQueryComboBoxConditions({
  resource,
  fieldName,
  collectionRelationships,
  treeData,
  subViewRelationship,
  relatedTable,
}: {
  readonly resource: SpecifyResource<AnySchema>;
  readonly fieldName: string;
  readonly treeData: QueryComboBoxTreeData | undefined;
  readonly collectionRelationships: CollectionRelationships | undefined;
  readonly relatedTable: SpecifyTable;
  readonly subViewRelationship: Relationship | undefined;
}): RA<SpecifyResource<SpQueryField>> {
  const fields: WritableArray<SpecifyResource<SpQueryField>> = [];
  const treeResource = toTreeTable(resource);
  if (typeof treeResource === 'object') {
    const tableId = resource.specifyTable.tableId;
    // Add not-a-descendant condition
    if (!treeResource.isNew())
      fields.push(
        QueryFieldSpec.fromStringId(`${tableId}..nodeNumber`, false)
          .toSpQueryField()
          .set('isDisplay', false)
          .set('isNot', true)
          .set('operStart', queryFieldFilters.between.id)
          .set(
            'startValue',
            [
              treeResource.get('nodeNumber') ?? '',
              treeResource.get('highestChildNodeNumber') ?? '',
            ].join(',')
          )
      );
    if (fieldName === 'parent') {
      // Add rank limits
      let nextRankId = 0;
      if (typeof treeData === 'object') {
        let rankIndex =
          treeData.treeRanks.findIndex(
            // "rankId" is the original value; not updated with unsaved changes
            ({ rankId }) => rankId === treeResource.get('rankId')
          ) + 1;
        if (rankIndex !== 0) {
          while (
            rankIndex < treeData.treeRanks.length &&
            !treeData.treeRanks[rankIndex].isEnforced
          )
            rankIndex += 1;
          nextRankId = treeData.treeRanks[rankIndex - 1].rankId;
        }
      }
      const lastTreeRankId = treeData?.treeRanks.at(-1)!.rankId ?? 0;

      const lowestRankId = Math.min(
        lastTreeRankId,
        nextRankId || lastTreeRankId,
        treeData?.lowestChildRank ?? lastTreeRankId
      );
      if (lowestRankId !== 0)
        fields.push(
          QueryFieldSpec.fromStringId(`${tableId}..rankId`, false)
            .toSpQueryField()
            .set('isDisplay', false)
            .set('startValue', lowestRankId.toString())
            .set('operStart', queryFieldFilters.less.id)
        );
    } else if (fieldName === 'acceptedParent') {
      // Nothing to do
    } else if (fieldName === 'hybridParent1' || fieldName === 'hybridParent2') {
      // Nothing to do
    }
  }

  if (
    typeof collectionRelationships === 'object' &&
    fieldName === 'collectionRelType'
  )
    // Add condition for current collection
    fields.push(
      QueryFieldSpec.fromStringId(
        `${relatedTable.tableId}..collectionRelTypeId`,
        true
      )
        .toSpQueryField()
        .set('isDisplay', false)
        .set('operStart', queryFieldFilters.in.id)
        .set(
          'startValue',
          collectionRelationships[
            subViewRelationship?.name === 'leftSideRels' ? 'left' : 'right'
          ]
            .map(({ id }) => id)
            .join(',')
        )
    );
  return fields;
}

export const getRelatedCollectionId = (
  { left, right }: CollectionRelationships,
  resource: SpecifyResource<AnySchema>,
  fieldName: string
): number | undefined =>
  (fieldName === 'rightSide'
    ? left
    : fieldName === 'leftSide'
    ? right
    : undefined
  )?.find(
    ({ id }) =>
      id ===
      idFromUrl(
        toTable(resource, 'CollectionRelationship')?.get('collectionRelType') ??
          ''
      )
  )?.collection;

/**
 * If some value is currently in the input field, try to figure out which
 * field it is intended for and populate that field in the new resource.
 * Most of the time, that field is determined based on the search field
 */
export function pendingValueToResource(
  relationship: Relationship,
  typeSearch: TypeSearch | false | undefined,
  pendingValue: string
): SpecifyResource<AnySchema> {
  const mainFields = getMainTableFields(relationship.relatedTable.name);
  const typeSearchFields =
    (typeof typeSearch === 'object'
      ? typeSearch?.searchFields
          .filter(
            ([searchField]) =>
              !searchField.isRelationship &&
              searchField.table === relationship.relatedTable
          )
          .map(([field]) => field)
      : undefined) ?? [];
  const fieldName = (
    mainFields.find((field) => typeSearchFields.includes(field)) ??
    mainFields[0]
  )?.name;
  return new relationship.relatedTable.Resource(
    typeof fieldName === 'string' ? { [fieldName]: pendingValue } : {}
  );
}
