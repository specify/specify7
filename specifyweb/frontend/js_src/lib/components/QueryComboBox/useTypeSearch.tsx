import React from 'react';

import { useAsyncState } from '../../hooks/useAsyncState';
import { getAppResourceUrl } from '../../utils/ajax/helpers';
import { filterArray } from '../../utils/types';
import type { LiteralField, Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { load } from '../InitialContext';
import { xmlToSpec } from '../Syncer/xmlUtils';
import type { TypeSearch } from './spec';
import { typeSearchesSpec } from './spec';

export const typeSearches = Promise.all([
  load<Element>(getAppResourceUrl('TypeSearches'), 'text/xml'),
  import('../DataModel/tables').then(async ({ fetchContext }) => fetchContext),
]).then(([xml]) =>
  filterArray(xmlToSpec(xml, typeSearchesSpec()).typeSearches)
);

export function useTypeSearch(
  initialTypeSearch: TypeSearch | string | undefined,
  field: LiteralField | Relationship,
  initialRelatedTable: SpecifyTable | undefined
): TypeSearch | false | undefined {
  const relatedTable =
    initialRelatedTable ??
    (field.isRelationship ? field.relatedTable : undefined) ??
    (typeof initialTypeSearch === 'object'
      ? initialTypeSearch.table
      : undefined);
  const [typeSearch] = useAsyncState<TypeSearch | false>(
    React.useCallback(async () => {
      if (typeof initialTypeSearch === 'object') return initialTypeSearch;
      else if (relatedTable === undefined) return false;
      return typeSearches.then(
        (typeSearches) =>
          typeSearches.find(({ name }) => name === initialTypeSearch) ??
          typeSearches.find(({ table }) => table === relatedTable) ??
          false
      );
    }, [initialTypeSearch, relatedTable]),
    false
  );
  return typeSearch;
}
