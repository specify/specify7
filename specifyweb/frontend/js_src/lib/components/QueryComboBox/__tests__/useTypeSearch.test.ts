import { requireContext } from '../../../tests/helpers';
import { typeSearches } from '../useTypeSearch';

requireContext();

test('type searches are fetched and parsed correctly', async () =>
  expect(
    typeSearches.then((result) =>
      // Remove symbols
      JSON.parse(JSON.stringify(result))
    )
  ).resolves.toMatchSnapshot());
