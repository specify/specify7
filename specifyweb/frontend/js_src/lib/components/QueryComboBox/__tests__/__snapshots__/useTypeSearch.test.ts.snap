// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type searches are fetched and parsed correctly 1`] = `
[
  {
    "displayFields": [
      "accessionNumber",
    ],
    "format": "%s",
    "name": "AccessionCO",
    "searchFields": [
      [
        "[literalField Accession.accessionNumber]",
      ],
    ],
    "table": "[table Accession]",
    "title": "Searched fields: Accession #",
  },
  {
    "displayFields": [
      "lastName",
      "firstName",
    ],
    "format": "%s, %s",
    "formatter": "Agent",
    "name": "Agent",
    "searchFields": [
      [
        "[literalField Agent.lastName]",
      ],
    ],
    "table": "[table Agent]",
    "title": "Searched fields: Last Name",
  },
  {
    "displayFields": [
      "lastName",
      "firstName",
    ],
    "format": "%s, %s",
    "formatter": "Agent",
    "name": "AgentSSPU",
    "searchFields": [
      [
        "[literalField Agent.lastName]",
      ],
    ],
    "table": "[table Agent]",
    "title": "Searched fields: Last Name",
  },
  {
    "displayFields": [
      "appraisalNumber",
    ],
    "format": "%s",
    "name": "Appraisal",
    "searchFields": [
      [
        "[literalField Appraisal.appraisalNumber]",
      ],
    ],
    "table": "[table Appraisal]",
    "title": "Searched fields: Appraisal Number",
  },
  {
    "displayFields": [
      "schemeName",
    ],
    "format": "%s",
    "name": "AutoNumberingScheme",
    "searchFields": [
      [
        "[literalField AutoNumberingScheme.schemeName]",
      ],
    ],
    "table": "[table AutoNumberingScheme]",
    "title": "Searched fields: Schema Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "name": "BioStrat",
    "searchFields": [
      [
        "[literalField GeologicTimePeriod.name]",
      ],
    ],
    "table": "[table GeologicTimePeriod]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "schemeName",
    ],
    "format": "%s",
    "name": "CatAutoNumberingScheme",
    "searchFields": [
      [
        "[literalField AutoNumberingScheme.schemeName]",
      ],
    ],
    "table": "[table AutoNumberingScheme]",
    "title": "Searched fields: Schema Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "name": "ChronosStrat",
    "searchFields": [
      [
        "[literalField GeologicTimePeriod.name]",
      ],
    ],
    "table": "[table GeologicTimePeriod]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "stationFieldNumber",
      "startDate",
      "loc.localityName",
      "geo.fullName",
    ],
    "format": "%s; %s; %s; %s",
    "name": "CollectingEvent",
    "searchFields": [
      [
        "[literalField CollectingEvent.stationFieldNumber]",
      ],
    ],
    "table": "[table CollectingEvent]",
    "title": "Searched fields: Field Number",
  },
  {
    "displayFields": [
      "text1",
      "text2",
      "number1",
    ],
    "format": "%s:%s:%s",
    "formatter": "CollectingTrip",
    "name": "CollectingTrip",
    "searchFields": [
      [
        "[literalField CollectingTrip.collectingTripName]",
      ],
    ],
    "table": "[table CollectingTrip]",
    "title": "Searched fields: Trip Name",
  },
  {
    "displayFields": [
      "catalogNumber",
    ],
    "format": "",
    "formatter": "CollectionObject",
    "name": "CollectionObject",
    "searchFields": [
      [
        "[literalField CollectionObject.catalogNumber]",
      ],
    ],
    "table": "[table CollectionObject]",
    "title": "Searched fields: Cat #",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "name": "Container",
    "searchFields": [
      [
        "[literalField Container.name]",
      ],
    ],
    "table": "[table Container]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "name": "ContainerOwner",
    "searchFields": [
      [
        "[literalField Container.name]",
      ],
    ],
    "table": "[table Container]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "primerDesignator",
    ],
    "format": "%s",
    "name": "DNAPrimer",
    "searchFields": [
      [
        "[literalField DNAPrimer.primerDesignator]",
      ],
    ],
    "table": "[table DNAPrimer]",
    "title": "Searched fields: Primer Designator",
  },
  {
    "displayFields": [
      "deaccessionNumber",
    ],
    "format": "%s",
    "formatter": "Deaccession",
    "name": "Deaccession",
    "searchFields": [
      [
        "[literalField Deaccession.deaccessionNumber]",
      ],
    ],
    "table": "[table Deaccession]",
    "title": "Searched fields: Deacc #",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "name": "Discipline",
    "searchFields": [
      [
        "[literalField Discipline.name]",
      ],
    ],
    "table": "[table Discipline]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "disposalNumber",
    ],
    "format": "%s",
    "formatter": "Disposal",
    "name": "Disposal",
    "searchFields": [
      [
        "[literalField Disposal.disposalNumber]",
      ],
    ],
    "table": "[table Disposal]",
    "title": "Searched fields: disposalNumber",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "name": "Division",
    "searchFields": [
      [
        "[literalField Division.name]",
      ],
    ],
    "table": "[table Division]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "title",
    ],
    "format": "%s",
    "name": "Exsiccata",
    "searchFields": [
      [
        "[literalField Exsiccata.title]",
      ],
    ],
    "table": "[table Exsiccata]",
    "title": "Searched fields: Title",
  },
  {
    "displayFields": [
      "fnbp.pageNumber",
      "fnb.name",
    ],
    "format": "%s - %s",
    "formatter": "FieldNotebookPage",
    "name": "FieldNotebookPage",
    "searchFields": [
      [
        "[literalField FieldNotebookPage.pageNumber]",
      ],
    ],
    "table": "[table FieldNotebookPage]",
    "title": "Searched fields: Page Number",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "Geography",
    "name": "Geography",
    "searchFields": [
      [
        "[literalField Geography.name]",
      ],
    ],
    "table": "[table Geography]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "name": "GeologicTimePeriod",
    "searchFields": [
      [
        "[literalField GeologicTimePeriod.name]",
      ],
    ],
    "table": "[table GeologicTimePeriod]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "journalName",
    ],
    "format": "%s",
    "name": "Journal",
    "searchFields": [
      [
        "[literalField Journal.journalName]",
      ],
    ],
    "table": "[table Journal]",
    "title": "Searched fields: Journal Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "name": "LithoStrat",
    "searchFields": [
      [
        "[literalField LithoStrat.name]",
      ],
    ],
    "table": "[table LithoStrat]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "localityName",
      "geo.fullName",
    ],
    "format": "%s;  %s",
    "formatter": "Locality",
    "name": "Locality",
    "searchFields": [
      [
        "[literalField Locality.localityName]",
      ],
    ],
    "table": "[table Locality]",
    "title": "Searched fields: Locality Name",
  },
  {
    "displayFields": [
      "viewName",
    ],
    "format": "%s",
    "name": "MorphBankView",
    "searchFields": [
      [
        "[literalField MorphBankView.viewName]",
      ],
    ],
    "table": "[table MorphBankView]",
    "title": "Searched fields: View Name",
  },
  {
    "displayFields": [
      "paleoContextName",
      "cs.fullName",
      "cse.fullName",
      "ls.fullName",
    ],
    "format": "%s, %s, %s, %s",
    "formatter": "PaleoContext",
    "name": "PaleoContext",
    "searchFields": [
      [
        "[literalField PaleoContext.paleoContextName]",
      ],
      [
        "[relationship PaleoContext.chronosStrat]",
        "[literalField GeologicTimePeriod.fullName]",
      ],
      [
        "[relationship PaleoContext.lithoStrat]",
        "[literalField LithoStrat.fullName]",
      ],
    ],
    "table": "[table PaleoContext]",
    "title": "Searched fields: Context Name, Chronostratigraphy / Full Name, and Lithostratigraphy / Full Name",
  },
  {
    "displayFields": [
      "permitNumber",
    ],
    "format": "%s",
    "name": "Permit",
    "searchFields": [
      [
        "[literalField Permit.permitNumber]",
      ],
    ],
    "table": "[table Permit]",
    "title": "Searched fields: Permit Number",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "name": "PrepType",
    "searchFields": [
      [
        "[literalField PrepType.name]",
      ],
    ],
    "table": "[table PrepType]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "title",
    ],
    "format": "%s",
    "name": "ReferenceWork",
    "searchFields": [
      [
        "[literalField ReferenceWork.title]",
      ],
    ],
    "table": "[table ReferenceWork]",
    "title": "Searched fields: Title",
  },
  {
    "displayFields": [
      "repositoryAgreementNumber",
    ],
    "format": "%s",
    "name": "RepositoryAgreement",
    "searchFields": [
      [
        "[literalField RepositoryAgreement.repositoryAgreementNumber]",
      ],
    ],
    "table": "[table RepositoryAgreement]",
    "title": "Searched fields: RA Number",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "name": "Storage",
    "searchFields": [
      [
        "[literalField Storage.name]",
      ],
    ],
    "table": "[table Storage]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "Taxon",
    "name": "Taxon",
    "searchFields": [
      [
        "[literalField Taxon.fullName]",
      ],
    ],
    "table": "[table Taxon]",
    "title": "Searched fields: Full Name",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "name": "TaxonTreeDefItem",
    "searchFields": [
      [
        "[literalField TaxonTreeDefItem.name]",
      ],
    ],
    "table": "[table TaxonTreeDefItem]",
    "title": "Searched fields: Name",
  },
]
`;
